gio/src/cancellable.rs

```diff
@@ -1,10 +1,153 @@
-    c.connect_cancelled(|_| {});
-    c.cancel(); // if it doesn't crash at this point, then we're good to go!
+// rustdoc-stripper-ignore-next
+/// The id of a cancelled handler that is returned by `CancellableExtManual::connect`. This type is
+/// analogous to `glib::SignalHandlerId`.
+#[derive(Debug, Eq, PartialEq)]
+#[repr(transparent)]
+pub struct CancelledHandlerId(libc::c_ulong);
+
+impl CancelledHandlerId {
+    // rustdoc-stripper-ignore-next
+    /// Returns the internal signal handler ID.
+    pub unsafe fn as_raw(&self) -> libc::c_ulong {
+        self.0
+    }
+}
+
+impl FromGlib<libc::c_ulong> for CancelledHandlerId {
+    #[inline]
+    unsafe fn from_glib(val: libc::c_ulong) -> Self {
+        Self(val)
+    }
+}
+
+pub trait CancellableExtManual {
+    #[doc(alias = "g_cancellable_connect")]
+    fn connect<F: FnOnce(&Self) + Send + Sync + 'static>(&self, f: F) -> CancelledHandlerId;
+    #[doc(alias = "g_cancellable_disconnect")]
+    fn disconnect(&self, id: CancelledHandlerId);
+    // rustdoc-stripper-ignore-next
+    /// Returns a `Future` that completes when the cancellable becomes cancelled. Completes
+    /// immediately if the cancellable is already cancelled.
+    fn future(&self) -> std::pin::Pin<Box<dyn Future<Output = ()> + Send + 'static>>;
+    #[doc(alias = "g_cancellable_source_new")]
+    fn source(&self) -> Source;
+}
+
+impl<O: IsA<Cancellable>> CancellableExtManual for O {
+    #[doc(alias = "g_cancellable_connect")]
+    fn connect<F: FnOnce(&Self) + Send + Sync + 'static>(&self, f: F) -> CancelledHandlerId {
+        unsafe extern "C" fn connect_trampoline<
+            P: IsA<Cancellable>,
+            F: FnOnce(&P) + Send + Sync + 'static,
+        >(
+            this: *mut ffi::GCancellable,
+            f: glib::ffi::gpointer,
+        ) {
+            let f: &mut Option<F> = &mut *(f as *mut Option<F>);
+            let f = f
+                .take()
+                .expect("Cancellable::cancel() closure called multiple times");
+            f(Cancellable::from_glib_borrow(this).unsafe_cast_ref())
+        }
+
+        unsafe extern "C" fn destroy_closure<F>(ptr: glib::ffi::gpointer) {
+            Box::<Option<F>>::from_raw(ptr as *mut _);
+        }
+
+        unsafe {
+            let f: Box<Option<F>> = Box::new(Some(f));
+            let id = ffi::g_cancellable_connect(
+                self.as_ptr() as *mut _,
+                Some(std::mem::transmute::<_, unsafe extern "C" fn()>(
+                    connect_trampoline::<Self, F> as *const (),
+                )),
+                Box::into_raw(f) as *mut _,
+                Some(destroy_closure::<F>),
+            );
+
+            from_glib(id)
+        }
+    }
+    #[doc(alias = "g_cancellable_disconnect")]
+    fn disconnect(&self, id: CancelledHandlerId) {
+        unsafe { ffi::g_cancellable_disconnect(self.as_ptr() as *mut _, id.as_raw()) };
+    }
+    fn future(&self) -> std::pin::Pin<Box<dyn Future<Output = ()> + Send + 'static>> {
```

Comment:
```suggestion
    fn future(&self) -> std::pin::Pin<Box<dyn Future<Output = ()> + Send + Sync + 'static>> {
```

Why is `Sync` not required?